source/extensions/filters/udp/udp_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,93 @@
+
+#include "envoy/extensions/filters/udp/udp_proxy/v3/route.pb.h"
+#include "envoy/extensions/filters/udp/udp_proxy/v3/route.pb.validate.h"
+#include "envoy/type/matcher/v3/network_inputs.pb.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/network/matching/data_impl.h"
+#include "source/common/network/matching/inputs.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace UdpProxy {
+namespace Router {
+
+Matcher::ActionFactoryCb RouteMatchActionFactory::createActionFactoryCb(
+    const Protobuf::Message& config, RouteActionContext& context,
+    ProtobufMessage::ValidationVisitor& validation_visitor) {
+  const auto& route_config = MessageUtil::downcastAndValidate<
+      const envoy::extensions::filters::udp::udp_proxy::v3::Route&>(config, validation_visitor);
+  const auto& cluster = route_config.cluster();
+
+  // Emplace cluster names to context to get all cluster names
+  context.cluster_name_.emplace(cluster);
+
+  return [cluster]() { return std::make_unique<RouteMatchAction>(cluster); };
+}
+
+REGISTER_FACTORY(RouteMatchActionFactory, Matcher::ActionFactory<RouteActionContext>);
+
+absl::Status RouteActionValidationVisitor::performDataInputValidation(
+    const Matcher::DataInputFactory<Network::NetworkMatchingData>&, absl::string_view type_url) {
+  static std::string source_ip_input_name = TypeUtil::descriptorFullNameToTypeUrl(
+      envoy::type::matcher::v3::SourceIpMatchInput::descriptor()->full_name());
+  if (type_url == source_ip_input_name) {
+    return absl::OkStatus();
+  }
+
+  return absl::InvalidArgumentError(
+      fmt::format("Route table can only match on source IP, saw {}", type_url));
+}
+
+RouterImpl::RouterImpl(const envoy::extensions::filters::udp::udp_proxy::v3::UdpProxyConfig& config,
+                       Server::Configuration::ServerFactoryContext& factory_context) {
+  if (config.has_cluster()) {
+    cluster_ = config.cluster();
+    cluster_names_.push_back(config.cluster());
+  } else {
+    RouteActionContext context{};
+    RouteActionValidationVisitor validation_visitor;
+    Matcher::MatchTreeFactory<Network::NetworkMatchingData, RouteActionContext> factory(
+        context, factory_context, validation_visitor);
+    matcher_ = factory.create(config.matcher())();
+
+    if (!validation_visitor.errors().empty()) {
+      // TODO(snowp): Output all violations.
+      throw EnvoyException(fmt::format("requirement violation while creating route match tree: {}",
+                                       validation_visitor.errors()[0]));
+    }
+
+    // Copy all clusters names
+    cluster_names_.insert(cluster_names_.end(), context.cluster_name_.begin(),
+                          context.cluster_name_.end());
+  }
+}
+
+const std::string RouterImpl::route(Network::Address::InstanceConstSharedPtr address) const {
+  if (cluster_.has_value()) {
+    return cluster_.value();
+  } else {
```

Comment:
nit: To remove a level of nesting in the code, since the body of the "if" returns, we can remove the "else" and out-dent everything below.

Reply:
Ok.